--- a/dwm.c	2013-02-21 22:43:33.000000000 +0100
+++ b/dwm.c	2013-03-01 16:52:04.339910568 +0100
@@ -2411,6 +2411,9 @@
 void
 view(const Arg *arg) {
 	unsigned int i;
+	int di;
+	unsigned int dui;
+	Window win, dummy;
 
 	if((arg->ui & TAGMASK) == selmon->tagset[selmon->seltags])
 		return;
@@ -2432,6 +2435,11 @@
 	selmon->lt[selmon->sellt]= selmon->lts[selmon->curtag];
 	focus(NULL);
 	arrange(selmon);
+
+	/* Focus window at current pointer location. */
+	XQueryPointer(dpy, root, &dummy, &win, &di, &di, &di, &di, &dui);
+	focus(wintoclient(win));
+	XFlush(dpy);
 }
 
 Client *
