--- dwm/dwm.c	2012-04-13 08:24:02.796424872 +0200
+++ dwm-dmenumon/dwm.c	2012-04-13 18:25:01.828621211 +0200
@@ -176,6 +176,7 @@
 static void detachstack(Client *c);
 static void die(const char *errstr, ...);
 static Monitor *dirtomon(int dir);
+static void dmenuspawn(const Arg *arg);
 static void drawbar(Monitor *m);
 static void drawbars(void);
 static void drawcoloredtext(Monitor *m, char *text);
@@ -868,6 +869,40 @@
 }
 
 void
+dmenuspawn(const Arg *arg) {
+   char monstr[2] = "0";
+   int len = 0;
+   int i;
+
+	const char** args = (const char**)(arg->v);
+   char** temp = (char**)(arg->v);
+
+   while (*temp++ != NULL)
+   	len++;
+   char **cmd = malloc((len + 3)*sizeof(char*));
+
+   for (i = 0; i < len; i++)
+   {
+   	cmd[i] = malloc((strlen(args[i]) + 1) * sizeof(char));
+   	strcpy(cmd[i], args[i]);
+   }
+
+   cmd[i] = malloc(3 * sizeof(char));
+   strcpy(cmd[i], "-m");
+   cmd[i + 1] = malloc(2 * sizeof(char));
+	monstr[0] = '0' + selmon->num;
+   strcpy(cmd[i + 1], monstr);
+   cmd[i + 2] = NULL;
+
+	Arg a = { .v = cmd };
+	spawn(&a);
+
+   for (i=0; i < len+2; i++)
+   	free(cmd[i]);
+   free(cmd);
+}
+
+void
 drawbar(Monitor *m) {
 	int x;
 	unsigned int i, occ = 0, urg = 0;
